File: numba/reduction.py
import time
import numpy as np
from numba import cuda

@cuda.jit
def sum_kernel(input_array, out):
    idx = cuda.threadIdx.x
    val = input_array[idx]
    shared = cuda.shared.array(8, dtype=np.float32)
    offset = 16
    while offset > 0:
        val += cuda.shfl_down_sync(0xffffffff, val, offset)
        offset //= 2
    if idx % 32 == 0:
        shared[idx >> 5] = val
    cuda.syncthreads()
    if idx <= 8:
        val = shared[idx]
        val += cuda.shfl_down_sync(0xffffffff, val, 4)
        val += cuda.shfl_down_sync(0xffffffff, val, 2)
        val += cuda.shfl_down_sync(0xffffffff, val, 1)
        if idx == 0:
            out[0] = val

@cuda.reduce
def sum_reduce(a, b):
    return a + b

def bench6():
    iterations_list = [1, 10, 100, 1000, 10000, 100000, 1000000]
    results = []
    size = 256
    pseudo_random_input = np.array([i * 123.0 for i in range(size)], dtype=np.float32)
    output = np.zeros(1, dtype=np.float32)
    expected = pseudo_random_input.sum()      # NumPy sum reduction

    # Precompile the kernels.
    d_temp1 = cuda.to_device(np.zeros(1, dtype=np.float32))
    d_temp2 = cuda.to_device(np.zeros(1, dtype=np.float32))
    sum_kernel[1, 1](d_temp1, d_temp2)
    cuda.synchronize()

    d_pseudo_random_input = cuda.to_device(pseudo_random_input)
    d_output = cuda.device_array_like(output)

    for iterations in iterations_list:
        # Reductions with Numba reduce:
        # start = time.perf_counter()
        # for _ in range(iterations):
        #     got = sum_reduce(pseudo_random_input)   # cuda sum reduction
        # elapsed = (time.perf_counter() - start) * 1000
        # print(f"bench6 - iterations: {iterations} | Time: {elapsed:.3f} ms")
        # print("Sum result:", got, ", expected:", expected)
        # results.append({
        #     'test': f'bench6_reduction',
        #     'size': iterations,
        #     'gpu_time_ms': elapsed
        # })
        start = time.perf_counter()
        for _ in range(iterations):
            sum_kernel[1, size](d_pseudo_random_input, d_output)
            cuda.synchronize()
        output = d_output.copy_to_host()
        elapsed = (time.perf_counter() - start) * 1000
        print(f"bench6 - iterations: {iterations} | Time: {elapsed:.3f} ms")
        print("Sum result:", output[0], ", expected:", expected)
        results.append({
            'test': f'bench6_sum_result: {output[0]}',
            'iterations': iterations,
            'gpu_time_ms': elapsed
        })
    return results

if __name__ == '__main__':
    bench6()

File: rust/src/reduction.rs
use cuda::dmem::{Buffer, DSend};
use cuda::{device_sync, gpu};
use crate::BenchResults;
use std::time::Instant;

#[kernel]
unsafe fn sum_kernel(input_array: &[f32], mut out: Buffer<f32>) {
    let idx = gpu::__nvvm_thread_idx_x() as usize;
    let val = input_array[idx];
// TODO: Add reduction implementation with shared memory
}

pub fn bench6() -> Vec<BenchResults> {
    match sum_kernel.pre_compile() {
        Ok(_) => {},
        Err(e) => {
            println!("Kernel compilation failed: {:?}", e);
            panic!("Kernel compilation failed");
        }
    }

    let mut results = Vec::new();

    let size = 256;
    let iterations_list = [1, 10, 100, 1000, 10000, 100000, 1000000];

    // Generate deterministic pseudo-random input
    let mut pseudo_random_input = vec![0f32; size];
    for i in 0..size {
        pseudo_random_input[i] = (i as f32 * 123.0) as f32;
    }
    let expect:f32 = pseudo_random_input.as_slice().iter().sum();

    let mut d_pseudo_random_input = pseudo_random_input.as_slice().to_device().unwrap();
    let mut d_out = Buffer::<f32>::alloc(1).unwrap().to_device().unwrap();

    for iterations in iterations_list {
        let start = Instant::now();
        for _ in 0..iterations {
            match sum_kernel.launch_with_dptr(
                size,
                1usize,
                &mut d_pseudo_random_input,
                &mut d_out,
            ) {
                Ok(_) => {}
                Err(e) => {
                    eprintln!("Error launching atomic_sum_kernel kernel: {:?}", e);
                }
            }
            match device_sync() {
                Ok(_) => {}
                Err(e) => {
                    eprintln!("Error synchronizing device after atomic_sum_kernel kernel: {:?}", e);
                }
            }
        }
        let out = d_out.retrieve().unwrap();
        let elapsed_gpu_ms = start.elapsed().as_secs_f64() * 1000.0;
        println!("for iterations {:?}, expected: {:?}, got: {:?}", iterations, expect, out.get(0));
        results.push(BenchResults {
            gpu_time: elapsed_gpu_ms,
            test: format!("bench6_sum_result: {:?}", out.get(0)),
            size: iterations,
        });
    }
    return results
}